Userland/Libraries/LibGfx/Path.cpp

```diff
@@ -393,136 +393,227 @@ struct RoundTrip {
-    constexpr auto flatness = 0.015f;
-    auto pen_vertex_count = max(
-        static_cast<int>(ceilf(AK::Pi<float> / acosf(1 - (2 * flatness) / thickness))), 4);
-    if (pen_vertex_count % 2 == 1)
-        pen_vertex_count += 1;
-
-    Vector<FloatPoint, 128> pen_vertices;
-    pen_vertices.ensure_capacity(pen_vertex_count);
-
-    // Generate vertices for the pen (going counterclockwise). The pen does not necessarily need
-    // to be a circle (or an approximation of one), but other shapes are untested.
-    float theta = 0;
-    float theta_delta = (AK::Pi<float> * 2) / pen_vertex_count;
-    for (int i = 0; i < pen_vertex_count; i++) {
-        float sin_theta;
-        float cos_theta;
-        AK::sincos(theta, sin_theta, cos_theta);
-        pen_vertices.unchecked_append({ cos_theta * thickness / 2, sin_theta * thickness / 2 });
-        theta -= theta_delta;
-    }
-
-    auto wrapping_index = [](auto& vertices, auto index) {
-        return vertices[(index + vertices.size()) % vertices.size()];
-    };
-
-    auto angle_between = [](auto p1, auto p2) {
-        auto delta = p2 - p1;
-        return atan2f(delta.y(), delta.x());
-    };
-
-    struct ActiveRange {
-        float start;
-        float end;
-
-        bool in_range(float angle) const
-        {
-            // Note: Since active ranges go counterclockwise start > end unless we wrap around at 180 degrees
-            return ((angle <= start && angle >= end)
-                || (start < end && angle <= start)
-                || (start < end && angle >= end));
-        }
-    };
+    Path stroked_path;
+    // 3. For each subpath of path:
+    for (auto& subpath : segments) {
+        // Let positions be the dash positions for the subpath.
+        float path_length = subpath.last().get<1>();
+        Vector<Tuple<float, float>> positions = compute_dash_positions(path_length, stroke_properties);
+        // NOTE: positions are guaranteed to be in ascending order
+        size_t start_index = 0;
+        size_t line_count = subpath.size() - 1;
+        // For each pair <start, end> in positions:
+        for (auto p : positions) {
+            float start = p.get<0>();
+            float end = min(p.get<1>(), path_length);
+            // Let dash be the shape that includes, for all distances between start and end along
+            // the subpath, all points that lie on the line perpendicular to the subpath at that
+            // distance and which are within distance ‘stroke-width’ of the point on the subpath at
+            // that position.
+            VERIFY(start - path_length < 1e-6f);
+            VERIFY(end - path_length < 1e-6f);
+            size_t line_index = start_index;
+            while (subpath[line_index + 1].get<1>() - start < 1e-6f)
+                line_index++;
+
+            Vector<NonnullRefPtr<Segment const>> out_path;
+            Vector<NonnullRefPtr<Segment const>> in_path;
```

Comment:
Is there a reason why these are not just Paths? Then everything below can use `in_path.append_segment<LineSegment>(...)` rather than the adopt_refs and such.

Reply:
The only reason these are not Paths is because at the end `in_path` has to be traversed backwards. I guess we can use `m_segments` at the end and use the better notation `in_path.append_segment<LineSegment>(...)`. I agree that it is very cumbersome but when writing it seemed like it was not that bad...